src/rc.rs

```diff
@@ -1894,15 +1894,20 @@ unsafe fn data_offset<T>(ptr: *const T) -> isize {
     (field_ptr - base_ptr) as isize
 }
 
+// Deallocate a `Box` without destroying the inner `T`.
+//
+// # Safety
+//
+// Callers must ensure that `ptr` was allocated by `Box::new` with the global allocator.
+//
+// Callers must ensure that `T` is not dropped.
 #[inline]
-// This signature has to be the same as `Box`, otherwise an ICE will happen.
-// When an additional parameter to `Box` is added (like `A: Allocator`), this has to be added here as
-// well.
-// For example if `Box` is changed to  `struct Box<T: ?Sized, A: Allocator>(Unique<T>, A)`,
-// this function has to be changed to `fn box_free<T: ?Sized, A: Allocator>(Unique<T>, A)` as well.
-pub(crate) unsafe fn box_free<T: ?Sized, A: Allocator>(ptr: NonNull<T>, alloc: A) {
-    let size = mem::size_of_val(ptr.as_ref());
-    let align = min_align_of_val(ptr.as_ref());
-    let layout = Layout::from_size_align_unchecked(size, align);
-    alloc.deallocate(ptr.cast(), layout);
+unsafe fn box_free<T: ?Sized, A: Allocator>(ptr: NonNull<T>, alloc: A) {
+    let _ignored = alloc;
+
+    // SAFETY: this function requires the allocated `T` behind `ptr` to not be
+    // dropped, so it is safe to convert the `NonNull<T>` to a `&T`.
+    let layout = Layout::for_value(ptr.as_ref());
```

Comment:
If `T` hasn't been dropped, this will fail to drop `T`, as you just call `dealloc`. Since you're on nightly anyway, you can (and probably should) use [`Layout::for_value_raw`](https://doc.rust-lang.org/stable/std/alloc/struct.Layout.html#method.for_value_raw) (which takes a raw pointer) instead.

Reply:
Not dropping `T` is intentional. This function is only called in `Rc::from_box` which copies the raw bytes of `T` out of the deallocated `Box`.

Reply:
I was originally using `Layout::for_value_raw` here but removed it to try and minimize the number of unstable features this crate relies on.

That was probably the wrong decision since avoiding turning a raw pointer to a reference here leaves less room for unsoundness.

This change should be made here and also in several places in `src/drop.rs`.

Thanks for the review!

Reply:
PR for switching to `Layout::for_value_raw` is here: https://github.com/artichoke/cactusref/pull/50

Thanks for the review!